### PR TITLE
#415: Adding check for controlling the max TCP connections

### PR DESCRIPTION
**Pull request by ayush571995**
For issue :
https://github.com/DiceDB/dice/issues/415

### CODE REVIEW

**Comment by ayush571995** [comment]
* The changes that are included in the PR.
  The change allows to limit the number of connections with the maxConnectionLimit

  * Design document, if any.

  * Information on any implementation choices that were made.

  * Evidence of sufficient testing. You ``MUST`` indicate the tests done, either manually or automated.
  Manual testing has been done. Ran the program with my changes and in local changed the maxConnectionLimit to 2 instead of 2000 from the config file. Created multiple connections from terminal and post the limit breach we are logging at server
```
        2024/08/30 20:51:43 INFO Password
        2024/08/30 20:51:43 INFO default user set up password required=false
        2024/08/30 20:52:16 WARN connection refused. Reached the max-connection limit
```
At client we close the connection with at terminal 
```
        truesaiyan@Asus-Rog-G15:~/oss/dicedb/dice$ dice -h localhost -p 7379
        Error while reading from localhost:7379 : (104, 'Connection reset by peer')
```

**Comment by JyotinderSingh** [comment]
Thanks for the PR, please add test cases verifying the functionality.

**Comment by ayush571995** [comment]
@JyotinderSingh The test cases are a little tricky here as I don't feel we should make 20k connections in integration test
These20k connections limit needs to be changed and would require change how we create async server . We need a new GetAsyncServer for testing and there we can hardcode the limit to some lower number and that is something which we should control with a flag 


So all these changes would be required

**Comment by JyotinderSingh** [comment]
> @JyotinderSingh The test cases are a little tricky here as I don't feel we should make 20k connections in integration test These20k connections limit needs to be changed and would require change how we create async server . We need a new GetAsyncServer for testing and there we can hardcode the limit to some lower number and that is something which we should control with a flag
> 
> So all these changes would be required

Yeah, that makes sense. We could launch a new server as a part of a separate test suite to allow ourselves to test such configuration-related behavior.

Additionally, we should be handling some additional edge cases, [similar to what redis does](https://redis.io/docs/latest/develop/reference/clients/#:~:text=Maximum%20Concurrent%20Connected%20Clients&text=The%20default%20is%2010%2C000%20clients,the%20soft%20limit%20is%20checked)

**Comment by lucifercr07** [comment]
@ayush571995 please rebase to resolve conflicts.

**Comment by lucifercr07** [comment]
@ayush571995 please check once the builds are failing.

**Comment by JyotinderSingh** [comment]
@ayush571995 Are you actively working on this?

**Comment by ayush571995** [comment]
> @ayush571995 Are you actively working on this?

Yes but I need some suggestions
How can we connect ?

**Comment by ayush571995** [comment]
@JyotinderSingh Added test can you please review ?

File: config/test_config.go
Comment (JyotinderSingh) [inline config/test_config.go:1]: We probably don't need this, you can set the config values as a part of the test.

File: internal/server/server.go [hunk 1]
```diff
@@ -55,6 +55,19 @@ func NewAsyncServer(shardManager *shard.ShardManager, watchChan chan dstore.Watc
...
+		maxClients:             config.TestServerMaxClients,
+		connectedClients:       make(map[int]*comm.Client),
+		shardManager:           shardManager,
+		queryWatcher:           querywatcher.NewQueryManager(),
+		multiplexerPollTimeout: config.TestServerMultiplexerPollTimeout,
+		ioChan:                 make(chan *ops.StoreResponse, 1000),
+		watchChan:              watchChan,
+	}
+}
+
```
Comment (JyotinderSingh) [inline internal/server/server.go:70]: We shouldn't need to define a new method like this.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for the changes. Left a few comments.

Also, we now have a server directory under `integration_tests/` directory. There you can create a new server and pass in the required config using the regular APIs. Please refer to the existing tests for this.

**Comment by lucifercr07** [comment]
@ayushsatyam146 please resolve the conflicts.

**Comment by ayushsatyam146** [comment]
I think you tagged the wrong Ayush @lucifercr07 😅

File: integration_tests/server/max_conn_test.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+func getConnection() (net.Conn, error) {
+	conn, err := net.Dial("tcp", fmt.Sprintf("localhost:%d", maxConnTestOptions.Port))
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+var maxConnTestOptions = commands.TestServerOptions{
```
[the remark this answers is not in the document]
Comment (ayush571995) [reply integration_tests/server/max_conn_test.go]: Done

File: integration_tests/server/max_conn_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	var maxConnLimit = config.TestServerMaxClients + 2
+	connections := make([]net.Conn, maxConnLimit)
+	for i := 0; i < maxConnLimit; i++ {
+		conn, err := getConnection()
+		if err == nil {
+			connections[i] = conn
+		} else {
+			t.Fatalf("unexpected error while getting connection %d: %v", i, err)
+		}
+		time.Sleep(500 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (ayush571995) [reply integration_tests/server/max_conn_test.go]: Removed

File: integration_tests/server/max_conn_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	_, err := getConnection()
+	assert.ErrorContains(t, err, "connect: connection refused")
+
+	result := commands.FireCommand(connections[0], "ABORT")
+	if result != "OK" {
+		t.Fatalf("Unexpected response to ABORT command: %v", result)
+	}
+
+	// Wait for the server to shut down
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (ayush571995) [reply integration_tests/server/max_conn_test.go]: Removed

File: config/test_config.go
Comment (ayush571995) [reply config/test_config.go:1]: Sure will do that

**Comment by ayush571995** [comment]
@JyotinderSingh @lucifercr07 
Made the changes can you guys have a look .

File: internal/server/server.go [hunk 5]
```diff
@@ -238,6 +242,10 @@ func (s *AsyncServer) eventLoop(ctx context.Context) error {
 
 // acceptConnection accepts a new client connection and subscribes to read events on the connection.
 func (s *AsyncServer) acceptConnection() error {
+	if len(s.connectedClients) > s.maxClients {
```
Comment (lucifercr07) [inline internal/server/server.go:245]: Ideally it's not a good practice have same number of maxClients value as the backlog queue max number which is being set during `listen()`. Shouldn't be a blocker for this PR, but we should revisit this later.

**Review by lucifercr07 — APPROVED** [review]
LGTM, thanks for contributing.
